archunit-example/example-junit4/src/test/java/com/tngtech/archunit/exampletest/junit4/ModulesTest.java

```diff
@@ -0,0 +1,174 @@
+
+    /**
+     * This example demonstrates how to check for cyclic dependencies between modules.
+     */
+    @ArchTest
+    public static ArchRule modules_should_be_free_of_cycles =
+            modules()
+                    .definedByAnnotation(AppModule.class)
+                    .should().beFreeOfCycles();
+
+    private static DescribedPredicate<ModuleDependency<AnnotationDescriptor<AppModule>>> declaredByDescriptorAnnotation() {
+        return DescribedPredicate.describe("declared by descriptor annotation", moduleDependency -> {
+            AppModule descriptor = moduleDependency.getOrigin().getDescriptor().getAnnotation();
+            List<String> allowedDependencies = stream(descriptor.allowedDependencies()).collect(toList());
+            return allowedDependencies.contains(moduleDependency.getTarget().getName());
+        });
+    }
+
+    private static IdentifierFromAnnotation identifierFromModulesAnnotation() {
+        return new IdentifierFromAnnotation();
+    }
+
+    private static DescriptorFunction<AnnotationDescriptor<AppModule>> fromModulesAnnotation() {
+        return DescriptorFunction.describe(String.format("from @%s(name)", AppModule.class.getSimpleName()),
+                (ArchModule.Identifier identifier, Set<JavaClass> containedClasses) -> {
+                    JavaClass rootClass = containedClasses.stream().filter(it -> it.isAnnotatedWith(AppModule.class)).findFirst().get();
+                    AppModule module = rootClass.getAnnotationOfType(AppModule.class);
+                    return new AnnotationDescriptor<>(module.name(), module);
+                });
+    }
+
+    private static class IdentifierFromAnnotation extends DescribedFunction<JavaClass, ArchModule.Identifier> {
+        IdentifierFromAnnotation() {
+            super("root classes with annotation @" + AppModule.class.getSimpleName());
+        }
+
+        @Override
+        public ArchModule.Identifier apply(JavaClass javaClass) {
+            return getIdentifierOfPackage(javaClass.getPackage());
+        }
+
+        private ArchModule.Identifier getIdentifierOfPackage(JavaPackage javaPackage) {
+            Optional<ArchModule.Identifier> identifierInCurrentPackage = javaPackage.getClasses().stream()
+                    .filter(it -> it.isAnnotatedWith(AppModule.class))
+                    .findFirst()
+                    .map(annotatedClassInPackage -> ArchModule.Identifier.from(annotatedClassInPackage.getAnnotationOfType(AppModule.class).name()));
+
+            return identifierInCurrentPackage.orElseGet(identifierInParentPackageOf(javaPackage));
+        }
+
+        private Supplier<ArchModule.Identifier> identifierInParentPackageOf(JavaPackage javaPackage) {
+            return () -> javaPackage.getParent().isPresent()
+                    ? getIdentifierOfPackage(javaPackage.getParent().get())
+                    : ArchModule.Identifier.ignore();
```

Comment:
Not relevant at all, but this could be:
```suggestion
            return () -> javaPackage.getParent()
                    .map(parent -> getIdentifierOfPackage(parent))
                    .orElseGet(ArchModule.Identifier::ignore);
```